cmd/fscrypt/setup.go

```diff
@@ -50,8 +50,22 @@ func createGlobalConfig(w io.Writer, path string) error {
 		return err
 	}
 
+	// v2 encryption policies are recommended, so set policy_version 2 when
+	// the kernel supports it. v2 policies are supported by upstream Linux
+	// v5.4 and later. For now we simply check the kernel version. Ideally
+	// we'd instead check whether setting a v2 policy actually works, in
+	// order to also detect backports of the kernel patches. However, that's
+	// hard because from this context (creating /etc/fscrypt.conf) we may
+	// not yet have access to a filesystem that supports encryption.
+	var policyVersion int64
+	if util.IsKernelVersionAtLeast(5, 4) {
```

Comment:
This is clever, but would it be possible to detect this by attempting to use one of the V2 ioctls?

I realize that `FS_IOC_GET_ENCRYPTION_POLICY_EX` can't distinguish between an old kernel and an FS with no encryption support (as they both return `ENOTTY`), but is the same true of `FS_IOC_ADD_ENCRYPTION_KEY`?

Does that return `ENOTTY` or `EOPNOTSUPP` on old kernels?

Also, is there kernel config for V2 policies (or a config that was added w/ V2 support). I found `CONFIG_FS_ENCRYPTION`, but that's for both V1 and V2 IIRC.

If neither of the above options work, what you have is acceptable.

Reply:
All the fscrypt ioctls fail with `ENOTTY` both on filesystem types that don't support encryption (e.g. "xfs" or "btrfs"), and on filesystem types that *do* support encryption ("ext4", "f2fs", and "ubifs") when the kernel is too old to know about the ioctl.

We could iterate through all mounted filesystems and hope there is one that supports encryption where we can check for the ioctl.  But the presence of such a filesystem during `fscrypt setup` is not guaranteed, so we'd have to fall back to checking the kernel version anyway.

Currently, v1 and v2 policies aren't configurable separately in the kernel config; there's just  `CONFIG_FS_ENCRYPTION=y`.  Also note it's not guaranteed that userspace has a way to check the kernel config anyway, since `/proc/config.gz` and `/boot/config-*` are not required.

So I think we should just go with the simple version number check for now.

Reply:
Version number check seems like our only option then, that's too bad.